rustls/tests/hpke.rs

```diff
@@ -0,0 +1,92 @@
+// There is presently no *ring* HPKE impl, and we can't easily make one ourselves because there's
+// no static ECDH. See https://github.com/briansmith/ring/issues/331
+#![cfg(feature = "aws_lc_rs")]
+
+use std::fs::File;
+
+use serde::Deserialize;
+
+use rustls::crypto::aws_lc_rs::hpke::DEFAULT_PROVIDER;
+use rustls::crypto::hpke::{HpkePrivateKey, HpkePublicKey, HpkeSuite};
+use rustls::internal::msgs::enums::{HpkeAead, HpkeKdf, HpkeKem};
+use rustls::internal::msgs::handshake::HpkeSymmetricCipherSuite;
+
+/// Confirm open/seal operations work using the test vectors from [RFC 9180 Appendix A].
+///
+/// [RFC 9180 Appendix A]: https://www.rfc-editor.org/rfc/rfc9180#TestVectors
+#[test]
+fn check_test_vectors() {
+    for (idx, vec) in test_vectors().into_iter().enumerate() {
+        if !vec.applicable() {
+            println!("skipping inapplicable vector {idx}");
+            continue;
+        }
+        println!("testing vector {idx}");
+
+        let mut hpke = DEFAULT_PROVIDER
+            .start(&vec.suite())
+            .unwrap();
+        let pk_r = HpkePublicKey(hex::decode(vec.pk_rm).unwrap());
+        let sk_r = HpkePrivateKey::from(hex::decode(vec.sk_rm).unwrap());
+        let info = hex::decode(vec.info).unwrap();
+
+        for enc in vec.encryptions {
+            let aad = hex::decode(enc.aad).unwrap();
+            let pt = hex::decode(enc.pt).unwrap();
+
+            let (enc, ciphertext) = hpke
+                .seal(&info, &aad, &pt, &pk_r)
+                .unwrap();
+
+            let plaintext = hpke
+                .open(&enc, &info, &aad, &ciphertext, &sk_r)
+                .unwrap();
+            assert_eq!(plaintext, pt);
+        }
+    }
+}
+
+#[derive(Deserialize, Debug)]
+struct TestVector {
```

Comment:
Not for this PR, but at some point I think we could make a "rustls-provider-test" crate that runs these and other known-answer and functional tests at the level of a `CryptoProvider`.